### ISSUE TITLE
Dot product between sparse vectors 2-4 times slower than naive implementation

### ISSUE THREAD

**Issue by tomtung** [vbarrielle/sprs#151]
I wrote a simple bench mark:
```rust
use rand::prelude::*;
use sprs::SpIndex;
use std::ops::AddAssign;

fn main() {
    const N_RUNS_PER_LOOP: usize = 10000;
    const N_LOOP: usize = 10;
    const MAX_DIM: usize = 100000;

    for _ in 0..N_LOOP {
        let dim = rand::thread_rng().gen_range(1, MAX_DIM);
        let vec1 = generate_random_csvec(dim);
        let vec2 = generate_random_csvec(dim);
        let expected = vec1.dot(&vec2);

        let start_t = time::precise_time_ns();
        for _ in 0..N_RUNS_PER_LOOP {
            assert_eq!(expected, vec1.dot(&vec2));
        }
        println!("CsVecBase::dot\t{}ns", time::precise_time_ns() - start_t);

        let start_t = time::precise_time_ns();
        for _ in 0..N_RUNS_PER_LOOP {
            assert_eq!(expected, csvec_dot(vec1.view(), vec2.view()));
        }
        println!("csvec_dot\t{}ns\n", time::precise_time_ns() - start_t);
    }
}

fn generate_random_csvec(dim: usize) -> sprs::CsVec<f32> {
    let mut rng = rand::thread_rng();
    let nnz = rng.gen_range(1, dim);

    let mut indices = rand::seq::index::sample(&mut rng, dim, nnz).into_vec();
    indices.sort();
    let data = (0..nnz).map(|_| rng.gen()).collect::<Vec<f32>>();

    sprs::CsVec::new(dim, indices, data)
}

pub fn csvec_dot<N, I>(
    vec1: sprs::CsVecViewI<N, I>,
    vec2: sprs::CsVecViewI<N, I>,
) -> N
    where
        I: SpIndex,
        N: num_traits::Num + Copy + AddAssign,
{
    let mut i = 0;
    let mut j = 0;
    let mut sum = N::zero();
    while i < vec1.nnz() && j < vec2.nnz() {
        let idx1 = vec1.indices()[i];
        let idx2 = vec2.indices()[j];
        if idx1 == idx2 {
            sum += vec1.data()[i] * vec2.data()[j];
        }

        if idx1 <= idx2 {
            i += 1;
        }

        if idx1 >= idx2 {
            j += 1;
        }
    }

    sum
}
```

One run prints the following:
```
CsVecBase::dot	8366680989ns
csvec_dot	2437925201ns

CsVecBase::dot	5487843624ns
csvec_dot	2000645107ns

CsVecBase::dot	2013433670ns
csvec_dot	666727261ns

CsVecBase::dot	8612510929ns
csvec_dot	3054450698ns

CsVecBase::dot	49202405ns
csvec_dot	14574851ns

CsVecBase::dot	1652846793ns
csvec_dot	564137218ns

CsVecBase::dot	446548963ns
csvec_dot	128573403ns

CsVecBase::dot	817144763ns
csvec_dot	259604342ns

CsVecBase::dot	3093001776ns
csvec_dot	1163095191ns

CsVecBase::dot	5155856292ns
csvec_dot	2091415900ns
```

The current implementation (`CsVecBase::dot`) is consistently much slower than a naive implementation (`csvec_dot`).

**Comment by vbarrielle** [comment]
The current dot product implementation is supposed to perform the exact same algorithm as your implementation, but I've over-engineered it a bit because I wanted to be able to call it on iterators...

I however thought the compiler would be able to see through the layers of abstraction and generate similar code, but clearly that is not the case here.

I does not seem impossible to write a version similar to yours based on iterators, so I should probably do that.

Thanks for the report!

**Comment by vbarrielle** [comment]
Fixed by #152
